Tests/SentryTests/Integrations/OutOfMemory/SentryOutOfMemoryScopeObserverTests.swift

```diff
@@ -0,0 +1,131 @@
+            breadcrumb.type = "user"
+            breadcrumb.message = "Click something"
+
+            options = Options()
+            fileManager = try! SentryFileManager(options: options, andCurrentDateProvider: currentDate)
+        }
+
+        func getSut() -> SentryOutOfMemoryScopeObserver {
+            return getSut(fileManager: self.fileManager)
+        }
+
+        func getSut(fileManager: SentryFileManager) -> SentryOutOfMemoryScopeObserver {
+            return SentryOutOfMemoryScopeObserver(maxBreadcrumbs: 10, fileManager: fileManager)
+        }
+    }
+
+    private var fixture: Fixture!
+    private var sut: SentryOutOfMemoryScopeObserver!
+
+    override func setUp() {
+        super.setUp()
+
+        fixture = Fixture()
+        sut = fixture.getSut()
+    }
+
+    override func tearDown() {
+        super.tearDown()
+        fixture.fileManager.deleteAllFolders()
+    }
+
+    // Test that we're storing the serialized breadcrumb in a proper JSON string
+    func testStoreBreadcrumb() throws {
+        let breadcrumb = fixture.breadcrumb.serialize() as! [String: String]
+
+        sut.addSerializedBreadcrumb(breadcrumb)
+
+        let fileOneContents = try String(contentsOfFile: fixture.fileManager.breadcrumbsFilePathOne)
+        let firstLine = String(fileOneContents.split(separator: "\n").first!)
+        let dict = try JSONSerialization.jsonObject(with: firstLine.data(using: .utf8)!) as! [String: String]
+
+        XCTAssertEqual(dict, breadcrumb)
+    }
+
+    func testStoreInMultipleFiles() throws {
+        let breadcrumb = fixture.breadcrumb.serialize()
+
+        var count = 0
+        while count < 9 {
+            sut.addSerializedBreadcrumb(breadcrumb)
+            count += 1
+        }
+
+        var fileOneContents = try String(contentsOfFile: fixture.fileManager.breadcrumbsFilePathOne)
+        var fileOnelines = fileOneContents.split(separator: "\n")
+        XCTAssertEqual(fileOnelines.count, 9)
+
+        XCTAssertFalse(FileManager.default.fileExists(atPath: fixture.fileManager.breadcrumbsFilePathTwo))
+
+        // Now store one more, which means it'll change over to the second file (which should be empty)
+        sut.addSerializedBreadcrumb(breadcrumb)
+
+        fileOneContents = try String(contentsOfFile: fixture.fileManager.breadcrumbsFilePathOne)
+        fileOnelines = fileOneContents.split(separator: "\n")
+        XCTAssertEqual(fileOnelines.count, 10)
+
+        var fileTwoContents = try String(contentsOfFile: fixture.fileManager.breadcrumbsFilePathTwo)
+        XCTAssertEqual(fileTwoContents, "")
+
+        // Next one will be stored in the second file
+        sut.addSerializedBreadcrumb(breadcrumb)
+
+        fileTwoContents = try String(contentsOfFile: fixture.fileManager.breadcrumbsFilePathTwo)
+        var fileTwolines = fileTwoContents.split(separator: "\n")
+
+        XCTAssertEqual(fileOnelines.count, 10)
+        XCTAssertEqual(fileTwolines.count, 1)
```

Comment:
`m`: If I set the max breadcrumbs to 10, I now have 11 when calling `SentryFileManager readPreviousBreadcrumbs`.

Reply:
I would drop them in `readPreviousBreadcrumbs`, as I could see people complaining or being confused about why they now have 199 breadcrumbs on an OOM.

Reply:
But the timestamp of a breadcrumb is optional, so we can't reliably sort them. And we don't know which of the two files has the older breadcrumbs either. Well we could see which one has the most breadcrumbs I guess, that should always be the older file. But it's starting to feel a bit iffy. Is that really worth it?

Reply:
We had issues in the past where people put a lot of data into breadcrumbs, so they reduce the maxBreadcrumbs to ensure not to exceed the max envelope item size. 

The initializer of the breadcrumb sets a `NSDate` https://github.com/getsentry/sentry-cocoa/blob/e2a3f3e232f56eca34f1c2f4f648d3c0d7ceccc9/Sources/Sentry/SentryBreadcrumb.m#L14

I think we can sort by the timestamp and if it's `nil` we consider it being older than the ones having timestamps. You have to manually set the timestamp to `nil`. I would sort them by timestamp and only send`maxBreadcrumbs`. I think the logic should not be in `readPreviousBreadcrumbs` as it's just reading them. I think it should be in `SentryOutOfMemoryTracker`.

Reply:
Sounds good, and thanks for giving the extra context about envelope size. 

Btw why is timestamp even nullable?